dango/app/src/proposal_preparer.rs

```diff
@@ -28,8 +34,85 @@ impl From<ProposerError> for grug_app::AppError {
     }
 }
 
-#[derive(Clone)]
-pub struct ProposalPreparer;
+pub struct ProposalPreparer {
+    params: Shared<Vec<(String, String)>>,
```

Comment:
Since the key is always `"ids[]"`, instead of `Vec<(String, String)>` we can do `Vec<&'static str, String>`? This avoids a heap allocation.